src/util/threadpool.rs

```diff
@@ -331,16 +341,17 @@ impl<Q, T> ThreadPool<Q, T>
             task_pool: task_pool,
             threads: threads,
             task_count: task_count,
+            sender: sender,
         }
     }
 
     pub fn execute<F>(&mut self, gid: T, job: F)
         where F: FnOnce() + Send + 'static
     {
-        let &(ref lock, ref cvar) = &*self.task_pool;
-        let mut meta = lock.lock().unwrap();
-        meta.push_task(gid, job);
+        let task = Task::new(gid, job);
+        self.sender.send(task).unwrap();
```

Comment:
In `endpoint` it would check the number of tasks in `threadpool` before push a new task into `pool`. Maybe we could add the limitation in another PR? @siddontang